File: Typecast/TCHmtxTable.m
//
//  TCHmtxTable.m
//  Typecast
//
//  Created by David Schweinsberg on 4/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCHmtxTable.h"
#import "TCDirectoryEntry.h"
#import "TCDataInput.h"
#import "TCHheaTable.h"
#import "TCMaxpTable.h"

@interface TCHmtxTable ()
{
    NSMutableArray *_hMetrics;
    NSMutableArray *_leftSideBearing;
}

@end


@implementation TCHmtxTable

- (id)initWithDataInput:(TCDataInput *)dataInput
         directoryEntry:(TCDirectoryEntry *)entry
              hheaTable:(TCHheaTable *)hhea
              maxpTable:(TCMaxpTable *)maxp
{
    self = [super init];
    if (self)
    {
        self.directoryEntry = [entry copy];
        _hMetrics = [[NSMutableArray alloc] initWithCapacity:[hhea numberOfHMetrics]];
        for (int i = 0; i < [hhea numberOfHMetrics]; ++i)
        {
            uint32_t hMetrics =
            [dataInput readUnsignedByte] << 24
            | [dataInput readUnsignedByte] << 16
            | [dataInput readUnsignedByte] << 8
            | [dataInput readUnsignedByte];
            [_hMetrics addObject:[NSNumber numberWithUnsignedInt:hMetrics]];
        }
        int lsbCount = [maxp numGlyphs] - [hhea numberOfHMetrics];
        _leftSideBearing = [[NSMutableArray alloc] initWithCapacity:lsbCount];
        for (int i = 0; i < lsbCount; ++i)
            [_leftSideBearing addObject:[NSNumber numberWithShort:[dataInput readShort]]];
    }
    return self;
}

- (uint16_t)advanceWidthAtIndex:(int)index
{
    if (_hMetrics == nil)
        return 0;
    if (index < [_hMetrics count])
        return [_hMetrics[index] unsignedIntValue] >> 16;
    else
        return [[_hMetrics lastObject] unsignedIntValue] >> 16;
}

- (int16_t)leftSideBearingAtIndex:(int)index
{
    if (_hMetrics == nil)
        return 0;
    if (index < [_hMetrics count])
        return (int16_t)([_hMetrics[index] unsignedIntValue] & 0xffff);
    else
        return [[_hMetrics lastObject] unsignedIntValue];
}

- (uint32_t)type
{
    return TCTable_hmtx;
}

- (NSString *)description
{
    NSMutableString *str = [NSMutableString stringWithFormat:
                            @"'hmtx' Table - Horizontal Metrics\n---------------------------------\n"
                            @"Size = %d bytes, %ld entries\n",
                            [[self directoryEntry] length],
                            [_hMetrics count]];
    for (int i = 0; i < [_hMetrics count]; ++i)
        [str appendFormat:
         @"        %d. advWid: %d, LSdBear: %d\n",
         i,
         [self advanceWidthAtIndex:i],
         [self leftSideBearingAtIndex:i]];

    for (int i = 0; i < [_leftSideBearing count]; ++i)
        [str appendFormat:
         @"        LSdBear %ld: %d\n",
         i + [_hMetrics count],
         [self leftSideBearingAtIndex:i]];
    return str;
}

@end

File: Typecast/TCHmtxTable.swift
//
//  TCHmtxTable.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/14/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

/**
 Horizontal Metrics
 */
class TCHmtxTable: TCBaseTable, Codable {
  var hMetrics: [UInt32] = []
  var leftSideBearings: [Int16] = []
  let dataCount: Int

  override init() {
    dataCount = 0
    super.init()
  }

  init(data: Data, hheaTable: TCHheaTable, maxpTable: TCMaxpTable) {
    dataCount = data.count
    let dataInput = TCDataInput(data: data)
    for _ in 0 ..< hheaTable.numberOfHMetrics {
      let v1 = UInt32(dataInput.readUInt8()) << 24
      let v2 = UInt32(dataInput.readUInt8()) << 16
      let v3 = UInt32(dataInput.readUInt8()) << 8
      let v4 = UInt32(dataInput.readUInt8())
      let metric = v1 | v2 | v3 | v4
      hMetrics.append(metric)
    }
    let lsbCount = maxpTable.numGlyphs - hheaTable.numberOfHMetrics
    for _ in 0..<lsbCount {
      leftSideBearings.append(dataInput.readInt16())
    }
    super.init()
  }

  func advanceWidth(at index: Int) -> Int {
    if index < hMetrics.count {
      return Int(hMetrics[index] >> 16)
    } else if let last = hMetrics.last {
      return Int(last >> 16)
    } else {
      return Int(hMetrics[hMetrics.count - 1] >> 16)
    }
  }

  func leftSideBearing(at index: Int) -> Int {
    if index < hMetrics.count {
      return Int(hMetrics[index] & 0xffff)
    } else if let last = hMetrics.last {
      return Int(last)
    } else {
      return Int(leftSideBearings[index - hMetrics.count])
    }
  }

  override class var tag: TCTableTag {
    get {
      return .hmtx
    }
  }

  override var description: String {
    get {
      var str = String.localizedStringWithFormat(
        "'hmtx' Table - Horizontal Metrics\n---------------------------------\n" +
        "Size = %d bytes, %ld entries\n",
        dataCount,
        hMetrics.count)
      for i in 0..<hMetrics.count {
        str.append("        \(i). advWid: \(advanceWidth(at: i)), LSdBear: \(leftSideBearing(at: i))\n")
      }
      for i in 0..<leftSideBearings.count {
        str.append("        LSdBear \(i + hMetrics.count): \(leftSideBearing(at: i))\n")
      }
      return str;
    }
  }
}
